test/Test/Util/FS.hs

```diff
@@ -347,6 +351,43 @@ listDirectoryFiles hfs = go Set.empty
   Corruption
 -------------------------------------------------------------------------------}
 
+-- | Flip a random bit in a random file in a given directory.
+flipRandomBitInRandomFile ::
+     (PrimMonad m, MonadThrow m)
+  => HasFS m h
+  -> Choice
+  -> FsPath
+  -> m (Maybe (FsPath, Int))
+flipRandomBitInRandomFile hfs bitChoice dir = do
+  -- List all files and their sizes
+  files <- fmap (dir </>) . toList <$> listDirectoryRecursiveFiles hfs dir
+  -- Handle the situation where there are no files
+  if null files then pure Nothing else do
+    filesAndFileSizeBits <-
+      for files $ \file -> do
+        fileSizeBytes <- withFile hfs file ReadMode (hGetSize hfs)
+        pure (file, fileSizeBytes * 8)
+    let totalFileSizeBits = sum (snd <$> filesAndFileSizeBits)
+    -- Handle the situation where there are no non-empty files
+    if totalFileSizeBits == 0 then pure Nothing else do
+      assert (totalFileSizeBits > 0) $ pure ()
+      -- Interpret `index` to point to a bit between `0` and `totalFileSize - 1`
+      let bitIndex = getChoice bitChoice (0, totalFileSizeBits - 1)
+      -- Flip the bit
+      Just <$> flipFileBitAt hfs bitIndex filesAndFileSizeBits
+
+-- | Internal helper: flip a single bit in a given list of files and sizes.
+flipFileBitAt :: (MonadThrow m, PrimMonad m) => HasFS m h -> Word64 -> [(FsPath, Word64)] -> m (FsPath, Int)
+flipFileBitAt _hfs bitIndex [] =
+  error $ printf "flipFileBitAt: bit index out of bounds (%d)" bitIndex
+flipFileBitAt hfs bitIndex ((file, fileSize) : filesAndSizes)
+  | bitIndex < fileSize = do
+    let bitIndexAsInt = fromIntegral bitIndex
```

Comment:
This could technically overflow (though highly unlikely). Maybe for correctness clamp the result to the `(0, maxBound)` range?